### ISSUE TITLE
List parameters are not encoded properly in Get request and Multipart/Form request even with ListFormat.multi (4.0.0-prev1)

### ISSUE THREAD

**Issue by lohanidamodar** [flutterchina/dio#1087]
In a Get and Multipart/Form request if any of the Parameter is a List and ListFormat is multi, previously in older versions the params were encoded properly (`[]`) was added in the params automatically. However with latest version, it doesn't get added. The culprit  is the following line
https://github.com/flutterchina/dio/blob/7abcfd1a758c45e81a11e7d63bb250ce8fe3fc25/dio/lib/src/utils.dart#L61

The check isCollection is false if a param is `z: ['string 1', 'string 2']`. The list is already checked in line
https://github.com/flutterchina/dio/blob/7abcfd1a758c45e81a11e7d63bb250ce8fe3fc25/dio/lib/src/utils.dart#L58
Is the check in `L61` necessary?

If it is, should we explicitly add `[]` to the parameter name in Get and Multipart/Form request?

Thank you for your help.

**Comment by wendux** [comment]
try `ListFormat.multiCompatible`

**Comment by lohanidamodar** [comment]
@wendux I tried both, its that line (L61) for both list types, that initiates the problem. No matter which format I use

**Comment by wendux** [comment]
1. L58 check `sub` , but L61  check `sub[i]`, they are different. 
 2. You'd better to give a case, and describe what you expected is，for example:
```dart 
  var data = {
    'a': '你好',
    'b': [5, '6'],
    'c': {
      'd': 8,
      'e': {
        'a': 5,
        'b': [66, 8]
      }
    }
  };
  test('#url encode multi', () {

   // a=你好&b[]=5&b[]=6&c[d]=8&c[e][a]=5&c[e][b][]=66&c[e][b][]=8
    var result =
        'a=%E4%BD%A0%E5%A5%BD&b%5B%5D=5&b%5B%5D=6&c%5Bd%5D=8&c%5Be%5D%5Ba%5D=5&c%5Be%5D%5Bb%5D%5B%5D=66&c%5Be%5D%5Bb%5D%5B%5D=8';
    expect(Transformer.urlEncodeMap(data, ListFormat.multiCompatible), result);
  });
```

**Comment by lohanidamodar** [comment]
Here is what I expect
```
 var data = {
    'a': 'string',
    'b': 'another_string',
    'z': ['string'],
  };
  test('#url encode multi', () {

   // a=string&b='another_string'&z[]=string
    var result =
        'a=string&b='another_string'&z%5B%5D=string';
    expect(Transformer.urlEncodeMap(data, ListFormat.multi), result);
  });

//Actual result I get: 'a=string&b='another_string'&z=string'
```

**Comment by barretzhi** [comment]
我也遇到这个问题，之前3.0没有问题

**Comment by wendux** [comment]
@lohanidamodar 

> Here is what I expect
> 
> ```
>  var data = {
>     'a': 'string',
>     'b': 'another_string',
>     'z': ['string'],
>   };
>   test('#url encode multi', () {
> 
>    // a=string&b=another_string&z[]=string
>     var result =
>         'a=string&b=another_string&z%5B%5D=string';
>     expect(Transformer.urlEncodeMap(data, ListFormat.multi), result);
>   });
> 
> //Actual result I get: 'a=string&b=another_string&z=string'
> ```

use `ListFormat.multiCompatible`, passing!

**Comment by lohanidamodar** [comment]
@wendux No matter what format I provide the encodeMap always receives the ListFormat.multi, and I believe the culprit here is this Compose function https://github.com/flutterchina/dio/blob/7534c1c9aa4577914a176e055a915b9b1220027a/dio/lib/src/options.dart#L274

If i however pass the listFormat to `RequestOptions` at line https://github.com/flutterchina/dio/blob/7abcfd1a758c45e81a11e7d63bb250ce8fe3fc25/dio/lib/src/options.dart#L270
 it works!!

**Comment by wendux** [comment]
> @wendux No matter what format I provide the encodeMap always receives the ListFormat.multi, and I believe the culprit here is this Compose function
> 
> https://github.com/flutterchina/dio/blob/7534c1c9aa4577914a176e055a915b9b1220027a/dio/lib/src/options.dart#L274
> 
> If i however pass the listFormat to `RequestOptions` at line
> 
> https://github.com/flutterchina/dio/blob/7abcfd1a758c45e81a11e7d63bb250ce8fe3fc25/dio/lib/src/options.dart#L270
> 
> 
> it works!!

fix in 4.0.0-prev3

**Comment by lohanidamodar** [comment]
Resolved!! ❤️ 👍🏻
